tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -720,31 +764,52 @@ def from_generator(generator, output_types, output_shapes=None, args=None):
         `iter()` protocol. If `args` is not specified, `generator` must take no
         arguments; otherwise it must take as many arguments as there are values
         in `args`.
-      output_types: A nested structure of `tf.DType` objects corresponding to
-        each component of an element yielded by `generator`.
+      output_types: (Optional.) A nested structure of `tf.DType` objects
+        corresponding to each component of an element yielded by `generator`.
+        Ignored if `output_spec` is specified.
       output_shapes: (Optional.) A nested structure of `tf.TensorShape` objects
         corresponding to each component of an element yielded by `generator`.
+        Ignored if `output_spec` is specified.
       args: (Optional.) A tuple of `tf.Tensor` objects that will be evaluated
         and passed to `generator` as NumPy-array arguments.
+      output_spec: (Optional.) A nested structure of `tf.TypeSpec` objects
+        corresponding to each component of an element yielded by `generator`.
 
     Returns:
       Dataset: A `Dataset`.
     """
     if not callable(generator):
       raise TypeError("`generator` must be callable.")
-    if output_shapes is None:
-      output_shapes = nest.map_structure(
+
+    if output_types is None and output_spec is None:
+      raise TypeError("Either `output_types` or `output_spec` must be "
+                      "specified.")
+
+    if output_spec is not None:
```

Comment:
We should also check that `output_shapes` is not set when `output_spec` is set.

Reply:
Right!

Reply:
done